File: csharp/1591.strange-printer-ii.cs
/*
 * @lc app=leetcode id=1591 lang=csharp
 *
 * [1591] Strange Printer II
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {

    private static bool HasCycle(int i, HashSet<int>[] deps, bool?[] visited) {
        visited[i] = false;
        foreach (int j in deps[i]) {
            if (visited[j].HasValue) {
                if (visited[j].Value) {
                    continue;
                }
                return true;
            }
            if (HasCycle(j, deps, visited)) {
                return true;
            }
        }
        visited[i] = true;
        return false;
    }

    public bool IsPrintable(int[][] targetGrid) {
        int m = targetGrid.Length, n = targetGrid[0].Length;
        int[][] boundaries = new int[61][];
        var deps = new HashSet<int>[61];
        var visited = new bool?[61];
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int c = targetGrid[i][j];
                if (boundaries[c] == null) {
                    boundaries[c] = new int[] { i, i, j, j };
                    continue;
                }
                boundaries[c][1] = i;
                boundaries[c][2] = Math.Min(boundaries[c][2], j);
                boundaries[c][3] = Math.Max(boundaries[c][3], j);
            }
        }
        for (int c = 1; c <= 60; ++c) {
            var boundary = boundaries[c];
            if (boundary == null) {
                continue;
            }
            var set = new HashSet<int>();
            for (int i = boundary[0]; i <= boundary[1]; ++i) {
                for (int j = boundary[2]; j <= boundary[3]; ++j) {
                    if (targetGrid[i][j] != c) {
                        set.Add(targetGrid[i][j]);
                    }
                }
            }
            deps[c] = set;
        }
        for (int c = 1; c <= 60; ++c) {
            if (deps[c] == null || visited[c].HasValue) {
                continue;
            }
            if (HasCycle(c, deps, visited)) {
                return false;
            }
        }
        return true;
    }
}
// @lc code=end

File: java/1591.strange-printer-ii.java
/*
 * @lc app=leetcode id=1591 lang=java
 *
 * [1591] Strange Printer II
 */

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

// @lc code=start

class Solution {
    public boolean isPrintable(int[][] targetGrid) {
        int m = targetGrid.length, n = targetGrid[0].length;
        Map<Integer, int[]> boundaries = new HashMap<>();
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int[] boundary = boundaries.get(targetGrid[i][j]);
                if (boundary == null) {
                    boundary = new int[] { i, i, j, j };
                    boundaries.put(targetGrid[i][j], boundary);
                    continue;
                }
                boundary[1] = i;
                boundary[2] = Math.min(boundary[2], j);
                boundary[3] = Math.max(boundary[3], j);
            }
        }
        Map<Integer, Set<Integer>> deps = new HashMap<>(boundaries.size());
        for (var colorToBoundary : boundaries.entrySet()) {
            int c = colorToBoundary.getKey();
            int[] boundary = colorToBoundary.getValue();
            Set<Integer> dep = new HashSet<>();
            for (int i = boundary[0]; i <= boundary[1]; ++i) {
                for (int j = boundary[2]; j <= boundary[3]; ++j) {
                    if (targetGrid[i][j] == c) {
                        continue;
                    }
                    dep.add(targetGrid[i][j]);
                }
            }
            deps.put(c, dep);
        }
        Map<Integer, Boolean> visited = new HashMap<>(boundaries.size());
        for (var c : boundaries.keySet()) {
            if (hasCycle(c, deps, visited)) {
                return false;
            }
        }
        return true;
    }

    private static boolean hasCycle(int curr, Map<Integer, Set<Integer>> deps, Map<Integer, Boolean> visited) {
        visited.put(curr, false);
        for (int next : deps.get(curr)) {
            Boolean v = visited.get(next);
            if (v != null) {
                if (v) {
                    continue;
                }
                return true;
            }
            if (hasCycle(next, deps, visited)) {
                return true;
            }
        }
        visited.put(curr, true);
        return false;
    }
}
// @lc code=end

